### PR TITLE
Fix FREELIST macros for AArch64

### PR DESCRIPTION
**Pull request by oknet**
This revert the commit 8563409 and fix a bug of FREELIST_VERSION(_x).

The layout of FREELIST_POINTER:

-  0 ~ 47 bit : 48 bits, Virtual Address
                (1+47 bits for AMD64 and 48 bits for AArch64)
- 48 ~ 62 bit : 15 bits, Freelist Version
- 63 bit :  1 bits, The type of Virtual Address
                (0 = user space, 1 = kernel space)

The valid Virtual Address range for AMD64 is:

- 0000000000000000 ~ 00007fffffffffff for user space
- ffff800000000000 ~ ffffffffffffffff for kernel space

The 47th bit can be used to indicate the type of VA.

And the one (64k page size) of the valid VA range for AArch64 is:

- 0000000000000000 ~ 0000ffffffffffff for user space
- ffff000000000000 ~ ffffffffffffffff for kernel space

We can see that the actual length of the Virtual Address is 48 bits on
AArch64 and 47 bits on AMD64. Therefore, we cannot use the 47th bit to
indicate the type of Virtual Address, which requires an extra bit.

The highest bit would be a good choice because we have done that before.

@dmorilha @rybit @biilmann @linnaea 

Link to issue #5016 .

### CODE REVIEW

**Comment by oknet** [comment]
The original design (before 85634092e22365b33a64655fb479be6b2ac8639b),

- 0 ~ 47 bit stores the VA address
- 48 ~ 62 bit stores the version
- 63 bit indicates the VA address is kernel space or user space


The `SET_FREELIST_POINTER_VERSION(_x,_p,_v)` store VA(_p) and VERSION(_v) into FREELIST_POINTER(_x):

- Each bit of High16(_p) has the same value, so it only needs to save 1 of the 16 bits, the highest bit.
- It save the highest 1 bit and the lowest 48 bits by `(((intptr_t)(_ p))) and 0x8000FFFFFFFFFFFFLL)`.
- And the rest 15 of 16 bits is used to save VERSION(_v) by `| (((_v)&0x7FFFLL) << 48)`

The `FREELIST_POINTER(_x)` return VA from a FREELIST_POINTER:

- if the VA address is from kernel space (start with 0xFFFFxxxxxxxxxxxx), the `FREELIST_POINTER(_x)` have to do 0xFFFF000000000000 | Low48(_x) and return, or just return Low48(_x).
- Low48(_x) = `(((intptr_t)(_x).data)&0xFFFFFFFFFFFFLL)` 
- 0xFFFF000000000000 or 0x0000000000000000 = `((((intptr_t)(_x).data)>>63)<<48)`
   - It fills the sign bit into all bits by right shift 63 times.
   - And clear the low 48 bits by left shift 48 times.

The `FREELIST_VERSION(_x)` returns VERSION from a FREELIST_POINTER, but there is a bug with `#define FREELIST_VERSION(_x) ((((intptr_t)(_x).data)<<1)>>49)`

- It tries to clear the highest bit by left shift 1 time, but it causes the sign bit filled by next bit.
- And then get the 15 bits VERSION by right shift 49 times.

We can just fix it simply by `#define FREELIST_VERSION(_x) ((((intptr_t)(_x).data)&0x7FFF000000000000LL)>>48)`:

- Just get 15 bits by `(((intptr_t)(_x).data)&0x7FFF000000000000LL)`
- And then get the 15 bits VERSION by right shift 48 times.

**Comment by SolidWallOfCode** [comment]
Is this related to #5444?

**Comment by SolidWallOfCode** [comment]
Whether right shift copies the sign bit is implementation dependent. It is legal for right shift to zero fill even on signed values. You can't guarantee that `(((intptr_t)(_x).data)>>63)<<48` copies the sign bit across the top 16 bits. That's why there's the extra bit of trickiness with subtracting 1 and bit flipping in the current version.

File: include/tscore/ink_queue.h [hunk 1]
```diff
@@ -137,10 +137,15 @@ union head_p {
...
 #elif defined(__x86_64__) || defined(__ia64__) || defined(__powerpc64__) || defined(__aarch64__) || defined(__mips64)
-#define FREELIST_POINTER(_x) \
-  ((void *)(((((intptr_t)(_x).data) << 16) >> 16) | (((~((((intptr_t)(_x).data) << 16 >> 63) - 1)) >> 48) << 48))) // sign extend
-#define FREELIST_VERSION(_x) (((intptr_t)(_x).data) >> 48)
-#define SET_FREELIST_POINTER_VERSION(_x, _p, _v) (_x).data = ((((intptr_t)(_p)) & 0x0000FFFFFFFFFFFFULL) | (((_v)&0xFFFFULL) << 48))
+/* Layout of FREELIST_POINTER
+ *
+ *  0 ~ 47 bits : 48 bits, Virtual Address (47 bits for AMD64 and 48 bits for AArch64)
+ * 48 ~ 62 bits : 15 bits, Freelist Version
+ *      63 bits :  1 bits, The type of Virtual Address (0 = user space, 1 = kernel space)
```
Comment (scw00) [inline include/tscore/ink_queue.h]: According to the memory map of AArch64, it seems the highest bit could be 1 in user space.

**Comment by oknet** [comment]
@SolidWallOfCode 

With the simple expression `((~0 >> 1)<0)`, we can detect which type of shift is implemented by the compiler.

**Comment by oknet** [comment]
More details about the virtual memory map in linux, please refer to:

- https://www.kernel.org/doc/Documentation/x86/x86_64/mm.txt
- https://www.kernel.org/doc/Documentation/arm64/memory.txt

**Comment by oknet** [comment]
@linnaea Could you please test this PR for issue #5016 ?

**Comment by dragon512** [comment]
I personally think this is not a smart idea. 

Just to be clear the Sunny Cove processor will have "Up to 4PB of memory is possible via 52-bit addresses. Virtual memory addresses expand up to 57 usable bits, allowing for 128PB"

I would feel confident that AMD is to follow this lead as well

**Comment by dragon512** [comment]
FYI, Masks will be faster here

File: include/tscore/ink_queue.h (hunk 1 above)
Comment (dmorilha) [reply include/tscore/ink_queue.h]: the explanation at the top looks great! thanks for that!

**Comment by dmorilha** [comment]
Thanks for giving attention to this.

In the end of the day I am more interested into making sure traffic server works in **aarch64** and less interested in the particulars of the implementation.

My personal taste is that playing with bits to save on space is ugly and leads to non maintainable code.

Another possible solution to this problem would be to have a different special version of the macros to `__aarch64__` (today we just bundle them all together).

I would also advise to play with alignment if possible rather than assuming the architecture addressable space. Here is how the DartVM distinguish things internally: https://github.com/dart-lang/sdk/blob/master/runtime/vm/pointer_tagging.h, they make sure pointers have the lower bits, rather than the higher ones, available for tagging.

File: include/tscore/ink_queue.h [hunk 2]
```diff
@@ -137,10 +137,28 @@ union head_p {
...
+#if ((~0 >> 1) < 0)
+/* the shift is `arithmetic' */
+#define FREELIST_POINTER(_x) ((void *)((((intptr_t)(_x).data) & 0x0000FFFFFFFFFFFFLL) | ((((intptr_t)(_x).data) >> 63) << 48)))
+#else
+/* the shift is `logical' */
 #define FREELIST_POINTER(_x) \
-  ((void *)(((((intptr_t)(_x).data) << 16) >> 16) | (((~((((intptr_t)(_x).data) << 16 >> 63) - 1)) >> 48) << 48))) // sign extend
-#define FREELIST_VERSION(_x) (((intptr_t)(_x).data) >> 48)
-#define SET_FREELIST_POINTER_VERSION(_x, _p, _v) (_x).data = ((((intptr_t)(_p)) & 0x0000FFFFFFFFFFFFULL) | (((_v)&0xFFFFULL) << 48))
+  ((void *)(((((intptr_t)(_x).data) << 16) >> 16) | (((~((((intptr_t)(_x).data) >> 63) - 1)) >> 48) << 48))) // sign extend
```
Comment (oknet) [inline include/tscore/ink_queue.h]: @dragon512 Do you mean the `((((intptr_t)(_x).data) << 16) >> 16)` is faster than `(((intptr_t)(_x).data) & 0x0000FFFFFFFFFFFFLL)` ?

@SolidWallOfCode This is different from what you told me in slack. I also think that the mask operation is faster.

**Comment by oknet** [comment]
@dragon512 It always use the bit 63 to indicate the type of Virtual Address, so it is easy to expand to 52-bit addresses by the similar code.

**Comment by linnaea** [comment]
@oknet Can confirm it works.

@dragon512 It's not a problem as these newer processors have 128-bit CAS, and unless the compiler is ancient, this code path won't get used.

**Comment by oknet** [comment]
It seems Masks more faster.

@SolidWallOfCode @dragon512 

```
bogon:eventsystem oknet$ gcc -O0 t.c -o t
bogon:eventsystem oknet$ ./t             
Test1: 95575811971446, time = 0.4851
Test2: 95575811971446, time = 0.3647
bogon:eventsystem oknet$ ./t
Test1: 95575811971446, time = 0.4832
Test2: 95575811971446, time = 0.3689
bogon:eventsystem oknet$ ./t
Test1: 95575811971446, time = 0.4447
Test2: 95575811971446, time = 0.3641
bogon:eventsystem oknet$ ./t
Test1: 95575811971446, time = 0.4679
Test2: 95575811971446, time = 0.3665
bogon:eventsystem oknet$ ./t
Test1: 95575811971446, time = 0.4721
Test2: 95575811971446, time = 0.3786
```

Test code:
```
#include <stdio.h>
#include <sys/time.h>

long int test1(long int in){
    int i;
    for(i = 0; i < 1000 * 1000; i++){
        in += i;
        in = ((((intptr_t)in) << 16) >> 16);
    }

    return i ^ in;
}

long int test2(long int in){
    int i;
    for(i = 0; i < 1000 * 1000; i++){
        in += i;
        in = (((intptr_t)in) & 0x0000FFFFFFFFFFFFLL);
    }

    return i ^ in;
}

int
main(int argc, char *argv[])
{
    struct timeval tv1, tv2;
    long int result;

    gettimeofday(&tv1, NULL);
    result = test1(0x1234567890123456LL);
    gettimeofday(&tv2, NULL);
    printf("Test1: %ld, time = %ld.%d\n", result, tv2.tv_sec - tv1.tv_sec, tv2.tv_usec - tv1.tv_usec);

    gettimeofday(&tv1, NULL);
    result = test2(0x1234567890123456LL);
    gettimeofday(&tv2, NULL);
    printf("Test2: %ld, time = %ld.%d\n", result, tv2.tv_sec - tv1.tv_sec, tv2.tv_usec - tv1.tv_usec);
}
```

**Pull request merged**